torch/csrc/jit/passes/onnx/peephole.cpp

```diff
@@ -830,6 +830,153 @@ void removeMaxPoolUnusedOutput(Block* b) {
+        origNllLossNode = *it;
+        deleteNodes.push_back(origLogSoftmaxNode);
+      // if the input is 4D
+      } else if (prev->kind() == onnx::Transpose &&
+                 prev->input(0)->node()->kind() == onnx::LogSoftmax) {
+        auto logSoftmaxNode = prev->input(0)->node();
+        origLogSoftmaxNode = logSoftmaxNode->input(0)->node();
+        origNllLossNode = *it;
+        prev->removeAllInputs();
+        logSoftmaxNode->removeAllInputs();
+        deleteNodes.push_back(origLogSoftmaxNode);
+        deleteNodes.push_back(logSoftmaxNode);
+        deleteNodes.push_back(prev);
+      // if the input is 3D or > 4D
+      } else if (prev->kind() == onnx::Reshape &&
+                 prev->input(0)->node()->kind() == onnx::Transpose &&
+                 prev->input(0)->node()->input(0)->node()->kind() == onnx::LogSoftmax) {
+        origLogSoftmaxNode = prev->input(0)->node()->input(0)->node();
+        auto transpose = origLogSoftmaxNode->input(0)->node();
+        origNllLossNode = *it;
+        auto reshape = origNllLossNode->input(1)->node();
+        std::vector<Value*> values{transpose->inputs().at(0), reshape->inputs().at(0)};
+        if (origNllLossNode->inputs().size() == 3) {
+          values.push_back(origNllLossNode->inputs().at(2));
+        }
+        softmaxCrossEntropyNode = createSoftmaxCrossEntropyNode(b, it, values);
+        origLogSoftmaxNode->removeAllInputs();
+        deleteNodes.push_back(origLogSoftmaxNode);
+        deleteNodes.push_back(transpose);
+        // when reduction=none a different graph is created and the graph doesn't end with
+        // node NegativeLogLikelihoodLoss like in all other cases and nodes that come after
+        // NegativeLogLikelihoodLoss need to be deleted.
+        if (!origNllLossNode->output(0)->uses().empty()){
+          auto nllloss_output = origNllLossNode->output(0)->uses()[0].user;
+ 
+          if (nllloss_output->kind() == onnx::Reshape &&
+              nllloss_output->inputs()[1]->node()->kind() == prim::ListConstruct) {
+            auto reshape_after = nllloss_output;
+            auto listconstruct = nllloss_output->inputs()[1]->node(); 
+ 
+            //make output of reshape the output of nllloss
+            reshape_after->replaceAllUsesWith(origNllLossNode);
+ 
+            reshape_after->removeAllInputs();
+            reshape_after->destroy();
+ 
+            auto gather = listconstruct->input(1)->node();
+  
+            listconstruct->removeAllInputs();
+            listconstruct->destroy();
+ 
+            auto shape = gather->input(0)->node();
+            auto constant = gather->input(1)->node();
+            constant->removeAllInputs();
+            shape->removeAllInputs();
+            gather->removeAllInputs();
+            constant->destroy();
+            shape->destroy();
+            gather->destroy();
+          } 
+        }
+
+        recursive_del(origNllLossNode);
+      } else {
+        break;
```

Comment:
break or continue?

Reply:
Maybe continue would be more correct, I put break to note that nll loss node doesn't belong to any of three cases but that would also break out of the for loop and not check if somewhere else in the graph there is nllloss not that does match one of the cases

Reply:
I think it should be continue too.